In [ ]:
# !pip install beautifulsoup4

# !pip install requests


from bs4 import BeautifulSoup
import requests
import pandas as pd
import numpy as np

In [ ]:
# For Never ending scraper


# url = "https://oldschool.runescape.wiki/w/Special:Random/main"
# url = "https://en.wikipedia.org/wiki/Category:Pages_containing_links_to_subscription-only_content?from=Le"
# url1 = "https://www.newegg.ca/abs-ali565/p/N82E16883360195?Item=N82E16883360195&cm_sp=Homepage_SS-_-P0_83-360-195-_-05072022"

# result = requests.get(url)

In [ ]:
# For Never ending scraper

# doc = BeautifulSoup(result.text, "html.parser")
# print(doc.prettify())

In [ ]:
# For Never ending scraper

# items = doc.find_all("b")
# print(items[0].parent)
# print(doc.body.text)

# title = doc.find(id="firstHeading")
# print(title.string)
# for item in items:
  # print(item.string)

In [ ]:
# For Never ending scraper


# import random

# # Get all the links
# allLinks = doc.find(id="bodyContent").find_all("a")
# # random.shuffle(allLinks)
# linkToScrape = 0
# print(allLinks)
# for link in allLinks:
#   print(link)
# 	# We are only interested in other wiki articles
#   # if link.find('href'):
#   if link.has_attr("href"):
#     if link['href'].find("/w/") == -1: 
#       continue
    
#   if not link.has_attr("href"):
#     continue

# 	# Use this link to scrape
#   linkToScrape = link
#   break

# # print(linkToScrape)

In [ ]:
# Never ending scraper

# def scrapeWikiArticle(url):
#   print(url)
#   response = requests.get(
#     url=url,
#   )
	
#   soup = BeautifulSoup(response.content, 'html.parser')

#   title = soup.find(id="firstHeading")
#   # print(url)
#   print(title.text)
#   allLinks = soup.find(id="bodyContent").find_all("a")
#   random.shuffle(allLinks)
#   linkToScrape = 0

#   for link in allLinks:
#     # We are only interested in other wiki articles
#     # print(link)
#     if link.has_attr("href"):
#       if link['href'].find("/wiki/") == -1: 
#         continue
    
#     if not link.has_attr("href"):
#       continue

#     if "http" in link['href']:
#       print("nepraeis")
#       continue
#     elif "//" in link['href']:
#       print("nepraeis")
#       continue
#     if ".org" in link['href']:
#       print("nepraeis")
#       continue
#     if ".com" in link['href']:
#       print("nepraeis")
#       continue
#     try: 
#       requests.get("https://en.wikipedia.org" + link['href'])
#     except Exception:
#       print("caught")
#       pass
#     # Use this link to scrape
#     # print(link)
#     linkToScrape = link
#     break
#   # print(linkToScrape['href'])
#   scrapeWikiArticle("https://en.wikipedia.org" + linkToScrape['href'])
# scrapeWikiArticle("https://en.wikipedia.org/wiki/RuneScape")

In [ ]:
# Wikipedia api scraper

!pip install wikipedia-api

In [ ]:
# Wikipedia api scraper

header_form = {"User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36", "Accept":"text/html,application/xhtml+xml,application/xml; q=0.9,image/webp,image/apng,/;q=0.8"}

In [ ]:
# Wikipedia api scraper

import wikipediaapi  # pip install wikipedia-api
import pandas as pd
import concurrent.futures
from tqdm import tqdm

def wiki_scrape(topic_name, verbose=True):
    def wiki_link(link):
        try:
            page = wiki_api.page(link)
            if page.exists():
                return {'page': link, 'text': page.text, 'link': page.fullurl,
                        'categories': list(page.categories.keys())}
        except:
            return None

  # Constructs Wikipedia object for extracting information Wikipedia.
    wiki_api = wikipediaapi.Wikipedia(language='en',
        extract_format=wikipediaapi.ExtractFormat.WIKI, headers=header_form)
  # Constructs Wikipedia page with title title.
    page_name = wiki_api.page(topic_name)
    if not page_name.exists():
        print('Page {} does not exist.'.format(topic_name))
        return
    
    # Returns all pages linked from the current page. tqdm - progress bar
    page_links = list(page_name.links.keys())
    progress = tqdm(desc='Links Scraped', unit='', total=len(page_links)) if verbose else None
    sources = [{'page': topic_name, 'text': page_name.text, 'link': page_name.fullurl,
                'categories': list(page_name.categories.keys())}]
    
    # The concurrent.futures module provides a high-level interface for asynchronously executing callables.
    with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
        future_link = {executor.submit(wiki_link, link): link for link in page_links}
        for future in concurrent.futures.as_completed(future_link):
            data = future.result()
            sources.append(data) if data else None
            progress.update(1) if verbose else None     
    progress.close() if verbose else None
    
    namespaces = ('Wikipedia', 'Special', 'Talk', 'LyricWiki', 'File', 'MediaWiki',
                  'Template', 'Help', 'User', 'Category talk', 'Portal talk')
    sources = pd.DataFrame(sources)
    sources = sources[(len(sources['text']) > 20)
                      & ~(sources['page'].str.startswith(namespaces, na=True))]
    sources['categories'] = sources.categories.apply(lambda x: [y[9:] for y in x])
    sources['topic'] = topic_name
    print('Wikipedia pages scraped:', len(sources))
    
    return sources

In [ ]:
# Wikipedia api scraper

def get_random_wiki_page_title(self):
  wiki_url_without_title = 'https://en.wikipedia.org/wiki/'
  random_page_url = requests.get('https://en.wikipedia.org/wiki/Special:Random').url
  return random_page_url.replace(wiki_url_without_title, "")

In [ ]:
# Wikipedia api scraper

title = get_random_wiki_page_title(0)

print(title)

Escadrille_62


In [ ]:
# pd_list = wiki_scrape("Black_hole")

In [ ]:
# len(pd_list.index._range)
# pd_list.shape[0]
# print(len(pd_list))

In [ ]:

# while (len(pd_list) < 10000):
#   title = get_random_wiki_page_title(0)
#   scraped = wiki_scrape(title)
#   pd_list = pd_list.append(scraped, ignore_index=True)
#   print(len(pd_list))  
# print(len(pd_list))

In [ ]:
# len(pd_list.index._range)
# pd_list.shape[0]

In [ ]:
# print(pd_list)

In [ ]:
# pd_list.to_csv('articles.txt', sep='t', index=False)

In [ ]:
# from pathlib import Path  
# filepath = Path('/content/drive/MyDrive/gmm/articles.txt') 

# pd_list.to_csv(filepath, sep='t', index=False)

# End of wiki scraper

In [ ]:
# texts = [pd_list['text'].loc[0], pd_list['text'].loc[1], pd_list['text'].loc[2], pd_list['text'].loc[1300]]

In [ ]:
!pip install -U sentence-transformers


     |████████████████████████████████| 79 kB 4.6 MB/s 
     |████████████████████████████████| 4.2 MB 18.3 MB/s 
     |████████████████████████████████| 1.2 MB 52.8 MB/s 
     |████████████████████████████████| 84 kB 1.9 MB/s 
     |████████████████████████████████| 596 kB 52.8 MB/s 
     |████████████████████████████████| 6.6 MB 35.4 MB/s 
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.0-py3-none-any.whl size=120747 sha256=13d615be13211ea17ffa2e238d8efe481dcfa790011713d0da4c37040f4e5715
  Stored in directory: /root/.cache/pip/wheels/83/c0/df/b6873ab7aac3f2465aa9144b6b4c41c4391cfecc027c8b07e7
Successfully built sentence-transformers
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
loc = '/content/drive/MyDrive/gmm/articles.txt'

In [ ]:
scraped_articles = pd.read_csv(loc, sep='t')


In [ ]:
scraped_articles

page  \
0                                         Black_hole   
1                                             3C 273   
2                                           A0620-00   
3                                 ADE classification   
4                                           ADM mass   
...                                              ...   
10207         Category:Articles with NTA identifiers   
10208       Category:Articles with SUDOC identifiers   
10209        Category:Articles with VIAF identifiers   
10210  Category:Articles with WORLDCATID identifiers   
10211       Category:Use dmy dates from October 2013   

                                                    text  \
0      A black hole is a region of spacetime where gr...   
1      3C 273 is a quasar located in the constellatio...   
2      A0620-00 (abbreviated from 1A 0620-00) is a bi...   
3      In mathematics, the ADE classification (origin...   
4      The ADM formalism (named for its authors Richa...   
...                                                  ...   
10207  Pages in this category should only be added by...   
10208  Pages in this category should only be added by...   
10209  Pages in this category should only be added by...   
10210  Pages in this category should only be added by...   
10211  Wikipedia articles (tagged in this month) that...   

                                                    link  \
0               https://en.wikipedia.org/wiki/Black_hole   
1                   https://en.wikipedia.org/wiki/3C_273   
2                 https://en.wikipedia.org/wiki/A0620-00   
3       https://en.wikipedia.org/wiki/ADE_classification   
4            https://en.wikipedia.org/wiki/ADM_formalism   
...                                                  ...   
10207  https://en.wikipedia.org/wiki/Category:Article...   
10208  https://en.wikipedia.org/wiki/Category:Article...   
10209  https://en.wikipedia.org/wiki/Category:Article...   
10210  https://en.wikipedia.org/wiki/Category:Article...   
10211  https://en.wikipedia.org/wiki/Category:Use_dmy...   

                                              categories            topic  
0      ['All articles containing potentially dated st...       Black_hole  
1      ['3C objects', 'Articles with short descriptio...       Black_hole  
2      ['Articles with short description', 'K-type ma...       Black_hole  
3      ['All articles with specifically marked weasel...       Black_hole  
4      ['Formalism (deductive)', 'Mathematical method...       Black_hole  
...                                                  ...              ...  
10207  ['Articles with authority control information'...  Usha_Kiran_Khan  
10208  ['Articles with authority control information'...  Usha_Kiran_Khan  
10209  ['Articles with authority control information'...  Usha_Kiran_Khan  
10210  ['Articles with authority control information'...  Usha_Kiran_Khan  
10211  ['CatAutoTOC generates Large category TOC', 'H...  Usha_Kiran_Khan  

[10212 rows x 5 columns]

In [ ]:
!pip install datasets

     |████████████████████████████████| 342 kB 5.9 MB/s 
     |████████████████████████████████| 1.1 MB 29.1 MB/s 
     |████████████████████████████████| 212 kB 41.9 MB/s 
     |████████████████████████████████| 136 kB 38.8 MB/s 
     |████████████████████████████████| 127 kB 43.5 MB/s 
     |████████████████████████████████| 271 kB 40.1 MB/s 
     |████████████████████████████████| 144 kB 29.7 MB/s 
     |████████████████████████████████| 94 kB 1.3 MB/s 
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.24.3
    Uninstalling urllib3-1.24.3:
      Successfully uninstalled urllib3-1.24.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.


In [ ]:
scraped_articles = scraped_articles[scraped_articles['text'].notnull()]
print(scraped_articles.iloc[737])

page                                Category:General relativity
text          General relativity is the best known relativis...
link          https://en.wikipedia.org/wiki/Category:General...
categories    ['CatAutoTOC generates standard Category TOC',...
topic                                                Black_hole
Name: 738, dtype: object


In [ ]:
from datasets import Dataset

articles_dataset = Dataset.from_pandas(scraped_articles)
articles_dataset

Dataset({
    features: ['page', 'text', 'link', 'categories', 'topic', '__index_level_0__'],
    num_rows: 10148
})

In [ ]:
print(articles_dataset['text'][737])

General relativity is the best known relativistic classical field theory of gravitation.


In [ ]:
articles_dataset = articles_dataset.map(
    lambda x: {"text_length": len(x["text"].split())}
)

  0%|          | 0/10148 [00:00<?, ?ex/s]

In [ ]:
articles_dataset = articles_dataset.filter(lambda x: x["text_length"] > 15)
articles_dataset

  0%|          | 0/11 [00:00<?, ?ba/s]

Dataset({
    features: ['page', 'text', 'link', 'categories', 'topic', '__index_level_0__', 'text_length'],
    num_rows: 9969
})

In [ ]:
articles_dataset = articles_dataset.filter(lambda x: x["text_length"] < 2000)
articles_dataset

  0%|          | 0/10 [00:00<?, ?ba/s]

Dataset({
    features: ['page', 'text', 'link', 'categories', 'topic', '__index_level_0__', 'text_length'],
    num_rows: 8100
})

In [ ]:
def concatenate_text(examples):
    return {
        "text": examples["topic"]
        + " \n "
        + examples["page"]
        + " \n "
        + examples["text"]
    }


articles_dataset = articles_dataset.map(concatenate_text)

  0%|          | 0/8100 [00:00<?, ?ex/s]

In [ ]:
texts = [scraped_articles['text'].loc[0], scraped_articles['text'].loc[1], scraped_articles['text'].loc[2], scraped_articles['text'].loc[1300]]

In [ ]:
scraped_articles.head(2)

page                                               text  \
0  Black_hole  A black hole is a region of spacetime where gr...   
1      3C 273  3C 273 is a quasar located in the constellatio...   

                                       link  \
0  https://en.wikipedia.org/wiki/Black_hole   
1      https://en.wikipedia.org/wiki/3C_273   

                                          categories       topic  
0  ['All articles containing potentially dated st...  Black_hole  
1  ['3C objects', 'Articles with short descriptio...  Black_hole

In [ ]:
# For longformer model

# from transformers import AutoTokenizer, AutoModel

# model_ckpt = "allenai/longformer-base-4096"
# tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
# model = AutoModel.from_pretrained(model_ckpt)


In [ ]:
# For bert model

from transformers import AutoTokenizer, AutoModel


model_ckpt = 'sentence-transformers/multi-qa-mpnet-base-dot-v1'

tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
model = AutoModel.from_pretrained(model_ckpt)

Downloading:   0%|          | 0.00/363 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/226k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/455k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/239 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/571 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/418M [00:00<?, ?B/s]

In [ ]:
# For GPU

import torch

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# device = torch.device("cuda")
model.to(device)

MPNetModel(
  (embeddings): MPNetEmbeddings(
    (word_embeddings): Embedding(30527, 768, padding_idx=1)
    (position_embeddings): Embedding(514, 768, padding_idx=1)
    (LayerNorm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): MPNetEncoder(
    (layer): ModuleList(
      (0): MPNetLayer(
        (attention): MPNetAttention(
          (attn): MPNetSelfAttention(
            (q): Linear(in_features=768, out_features=768, bias=True)
            (k): Linear(in_features=768, out_features=768, bias=True)
            (v): Linear(in_features=768, out_features=768, bias=True)
            (o): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (LayerNorm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (intermediate): MPNetIntermediate(
          (dense): Linear(in_features

In [ ]:
# as a single vector, so we need to “pool” or average our token embeddings in some way. 
# One popular approach is to perform CLS pooling on our 
# model’s outputs, where we simply collect the last hidden state
# for the special [CLS] token.

def cls_pooling(model_output):
    return model_output.last_hidden_state[:, 0]

In [ ]:
# For CPU

# Next, we’ll create a helper function that will 
# tokenize a list of documents, place the tensors on the GPU, 
# feed them to the model, and finally apply CLS pooling to the outputs:

# def get_embeddings(text_list):
#     encoded_input = tokenizer(
#         text_list, padding=True, truncation=True, return_tensors="pt"
#     )
#     encoded_input = {k: v for k, v in encoded_input.items()} #v.to(device)
#     model_output = model(**encoded_input)
#     return cls_pooling(model_output)

In [ ]:
# For GPU

# Tokenization is essentially splitting a phrase, sentence, paragraph,
# or an entire text document into smaller units, such as individual words or terms. 
# Each of these smaller units are called tokens.

# Next, we’ll create a helper function that will 
# tokenize a list of documents, place the tensors on the GPU, 
# feed them to the model, and finally apply CLS pooling to the outputs:

def get_embeddings(text_list):
    encoded_input = tokenizer(
        text_list, padding=True, truncation=True, return_tensors="pt"
    )
    encoded_input = {k: v.to(device) for k, v in encoded_input.items()} #v.to(device)
    model_output = model(**encoded_input)
    return cls_pooling(model_output)

In [ ]:
embedding = get_embeddings(articles_dataset["text"][0])
embedding.shape

torch.Size([1, 768])

Great, we’ve converted the first entry in our corpus into a 768-dimensional vector! We can use Dataset.map() to apply our get_embeddings() function to each row in our corpus, so let’s create a new embeddings column as follows:

Notice that we’ve converted the embeddings to NumPy arrays — that’s because 🤗 Datasets requires this format when we try to index them with FAISS, which we’ll do next.

In [ ]:
# embeddings_dataset = articles_dataset.map(
#     lambda x: {"embeddings": get_embeddings(x["text"]).detach().cpu().numpy()[0]}
# )

In [ ]:
# !pip uninstall faiss-gpu

!pip install faiss-gpu
!pip install faiss-cpu

     |████████████████████████████████| 85.5 MB 110 kB/s 
     |████████████████████████████████| 8.6 MB 8.7 MB/s 


In [ ]:
!pip install --upgrade pip

     |████████████████████████████████| 2.1 MB 7.2 MB/s 
  Attempting uninstall: pip
    Found existing installation: pip 21.1.3
    Uninstalling pip-21.1.3:
      Successfully uninstalled pip-21.1.3


Faiss is a library for efficient similarity search and clustering of dense vectors. It contains algorithms that search in sets of vectors of any size, up to ones that possibly do not fit in RAM. It also contains supporting code for evaluation and parameter tuning.

Faiss is written in C++ with complete wrappers for Python. Some of the most useful algorithms are implemented on the GPU. It is developed by Facebook AI Research.

What is similarity search?
Given a set of vectors x(i) in dimension d, Faiss build a data structure in RAM from it. After the structure is constructed, when given a new vector x in dimension d it performs efficiently the operation: j = argmin(i)|| x - x(i) ||

In [ ]:
import faiss

Now that we have a dataset of embeddings, we need some way to search over them. To do this, we’ll use a special data structure in 🤗 Datasets called a FAISS index. FAISS (short for Facebook AI Similarity Search) is a library that provides efficient algorithms to quickly search and cluster embedding vectors.

The basic idea behind FAISS is to create a special data structure called an index that allows one to find which embeddings are similar to an input embedding. Creating a FAISS index in 🤗 Datasets is simple — we use the Dataset.add_faiss_index() function and specify which column of our dataset we’d like to index:

In [ ]:
# embeddings_dataset.add_faiss_index(column="embeddings")

In [ ]:
import pickle


In [ ]:
# Longformer embeddings save

# with open('/content/drive/MyDrive/gmm/embeddings_dataset.pickle', 'wb') as f:
#   pickle.dump(embeddings_dataset, f)

In [ ]:
# Bert embeddings save

# with open('/content/drive/MyDrive/gmm/embeddings_dataset_bert.pickle', 'wb') as f:
#   pickle.dump(embeddings_dataset, f)

In [ ]:
# Longformer embeddings load

# with open('/content/drive/MyDrive/gmm/embeddings_dataset.pickle', 'rb') as f:
# 	loaded_dataset = pickle.load(f)

In [ ]:
# Bert embeddings load

with open('/content/drive/MyDrive/gmm/embeddings_dataset_bert.pickle', 'rb') as f:
  loaded_dataset = pickle.load(f) 
	# loaded_dataset_bert = pickle.load(f)

In [ ]:
loaded_dataset['text'][1]

'Black_hole \n A0620-00 \n A0620-00 (abbreviated from 1A 0620-00) is a binary star system in the constellation of Monoceros.\nA0620-00 consists of two objects. The first object is a K-type main-sequence star with a spectral type of K5 V. The second object cannot be seen, but based on its calculated mass of 6.6 M☉, it is too massive to be a neutron star and must therefore be a stellar-mass black hole. At a distance of about roughly 3,300 light-years (1,000 parsecs) away, this would make A0620-00 one of the nearest known black holes to the Solar System, closer than GRO J1655-40. The two objects orbit each other every 7.75 hours.\n\nA0620-00 has undergone two X-ray outbreaks. The first one was in 1917. The second burst, in 1975, was detected by the Ariel 5 satellite. During that time, A0620-00 was the brightest X-ray point source. It is now classified as an X-ray nova. Its black hole nature was determined in 1986.The black hole in A0620-00 pulls matter from the K-type star into an accreti

In [ ]:
# for i in loaded_dataset:
#   if i['topic'] == "Black_hole":
#     print(i['page'])

# loaded_dataset.drop_index('embeddings')

# new_dataset = loaded_dataset.filter(lambda x: x["topic"] == "Black_hole")

# loaded_dataset.add_faiss_index(column="embeddings")


# print(new_dataset)

In [ ]:
# print(new_dataset['page'])

We can now perform queries on this index by doing a nearest neighbor lookup with the Dataset.get_nearest_examples() function. Let’s test this out by first embedding a question as follows:

In [ ]:
question = "Frodo recovers in Rivendell under Elrond's care. The Council of Elrond discusses the history of Sauron and the Ring. Strider is revealed to be Aragorn, the heir of Isildur. Isildur had cut the One Ring from Sauron's hand in the battle ending the Second Age, but refused to destroy it, claiming it for himself. The Ring had been lost when Isildur was killed, finally ending up in Bilbo's possession, after his meeting with Gollum, described in The Hobbit. Gandalf reports that the chief wizard, Saruman, has betrayed them and is now working to become a power in his own right. Gandalf was captured by Saruman, but escaped, explaining why he had failed to return to meet Frodo as he had promised."
question_embedding = get_embeddings([question]).cpu().detach().numpy()
question_embedding.shape

(1, 768)

In [ ]:
question1 = "Text about korea "
question_embedding1 = get_embeddings([question1]).cpu().detach().numpy()
question_embedding1.shape

(1, 768)

Just like with the documents, we now have a 768-dimensional vector representing the query, which we can compare against the whole corpus to find the most similar embeddings:

In [ ]:
scores, samples = loaded_dataset.get_nearest_examples(
    "embeddings", question_embedding, k=len(loaded_dataset)
)

The Dataset.get_nearest_examples() function returns a tuple of scores that rank the overlap between the query and the document, and a corresponding set of samples (here, the 5 best matches). Let’s collect these in a pandas.DataFrame so we can easily sort them:

In [ ]:
import pandas as pd

samples_df = pd.DataFrame.from_dict(samples)
samples_df["scores"] = scores
samples_df.sort_values("scores", ascending=False, inplace=True)

In [ ]:
for i, row in samples_df.iterrows():
  if (i == 0 or i == 1 or i == 2 or i == 3 or i == 4 or i == len(samples_df) - 1):
    print(f"INDEX: {i}")
    print(f"TEXT: {row.text}")
    print(f"SCORE: {row.scores}")
    print(f"Topic: {row.topic}")
    print(f"LINK: {row.link}")
    print("=" * 50)
    print()

INDEX: 8099
TEXT: Syngenta 
 Knockout mouse 
 A knockout mouse, or knock-out mouse, is a genetically modified mouse (Mus musculus) in which researchers have inactivated, or "knocked out", an existing gene by replacing it or disrupting it with an artificial piece of DNA. They are important animal models for studying the role of genes which have been sequenced but whose functions have not been determined. By causing a specific gene to be inactive in the mouse, and observing any differences from normal behaviour or physiology, researchers can infer its probable function.
Mice are currently the laboratory animal species most closely related to humans for which the knockout technique can easily be applied. They are widely used in knockout experiments, especially those investigating genetic questions that relate to human physiology. Gene knockout in rats is much harder and has only been possible since 2003.The first recorded knockout mouse was created by Mario R. Capecchi, Martin Evans, and 

In [ ]:
print(loaded_dataset)

Dataset({
    features: ['page', 'text', 'link', 'categories', 'topic', '__index_level_0__', 'text_length', 'embeddings'],
    num_rows: 8100
})


In [ ]:
# print(loaded_dataset['text'][1])
# print(loaded_dataset['topic'][1])
# print(loaded_dataset['page'][1])

In [ ]:
# from sentence_transformers import SentenceTransformer
# from tqdm import tqdm
# from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
# from sentence_transformers import models


# word_embedding_model = models.Transformer('allenai/longformer-base-4096')
# pooling_model = models.Pooling(word_embedding_model.get_word_embedding_dimension())
# model = SentenceTransformer(modules=[word_embedding_model, pooling_model])

In [ ]:
# text_embeddings = model.encode(texts, batch_size = 4, show_progress_bar = True)

In [ ]:
# # doing the tokenization to prep data as tokens
# preprocessed_data = []

# for t in scraped_articles['text']:
#   encoded_input = tokenizer(t, return_tensors='pt')
#   preprocessed_data.append(encoded_input)

In [ ]:
# calculating last hidden states and everaging to have fixed dimmention features for each text
# outputs = []
# counter = 0
# for inputs in preprocessed_data:
#   counter += 1
#   output = model(**inputs)
#   features = np.mean(output[0].detach().numpy(), axis = 1)
#   outputs.append(features)

#   if counter > 100:
#     break

In [ ]:
# text_embeddings = model.encode(scraped_articles['text'], batch_size = 100, show_progress_bar = True)


In [ ]:
# np.shape(text_embeddings)

# similarities = cosine_similarity(text_embeddings)
# print('pairwise dense output:\n {}\n'.format(similarities))

# similarities_sorted = similarities.argsort()
# similarities_sorted

In [ ]:
# Get list of similarity indices i.e. doc at index 0 simialr with doc at index 1169 below.
# id_1 = []
# id_2 = []
# score = []
# for index,array in enumerate(similarities_sorted):
#     id_1.append(index)
#     id_2.append(array[-2])
#     score.append(similarities[index][array[-2]])
# index_df = pd.DataFrame({'id_1' : id_1,
#                           'id_2' : id_2,
#                           'score' : score})

In [ ]:
# index_df

In [ ]:
print(loaded_dataset.data)

InMemoryTable
page: string
text: string
link: string
categories: string
topic: string
__index_level_0__: int64
text_length: int64
embeddings: list<item: float>
  child 0, item: float
----
page: [["3C 273","A0620-00","ADE classification","ADM mass","ADM formalism","APM 08279+5255","Abraham H. Taub","A K Peters, Ltd.","Alan Guth","Alexander Friedmann",...,"Mars Surveyor 2001","Mars Telecommunications Orbiter","Mars Surveyor 2001 Lander","Mars aircraft","Mars flyby","Mars race","Mars rover","Marsokhod","Martian Moons Exploration","Medusae Fossae"],["Mercury Control Center","Meteor-3M No.1","Microwave limb sounder","Mission patch","Monkeys and apes in space","Mystic Mountain","Moon tree","Multi-Purpose Logistics Module","Multi-angle imaging spectroradiometer","Mythodea",...,"Ranjitpur","Rasulpur, Bachhrawan","Rasigaon","Rashidpur, Raebareli","Rastamau","Rasulpur, Amawan","Rasulpur, Rohaniya","Riwan","Rasulpur, Sareni","Rohaniya"],["Ratason","Reoli, Raebareli","Ratanpur, Raebareli","Rokha",